File: problems/0104-maximum-depth-of-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        return solve(root, 0);
    }

    int solve(TreeNode* root, int sum) {
        if (!root) return sum;

        int l = solve(root->left, sum + 1);
        int r = solve(root->right, sum + 1);

        if (l > r) return l;
        return r;
    }
};

File: problems/0104-maximum-depth-of-binary-tree/solution.py
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        return self.solve(root, 0)
    
    def solve(self, root, level):
        if not root: return level
        return max(self.solve(root.left, level+1), self.solve(root.right, level+1))
